### ISSUE TITLE
Distorted images when scanning below 200dpi (preview also distorted)

### ISSUE THREAD

**Issue by WayneManion** [sbs20/scanservjs#111]
I get garbled output when scanning below 200 dpi. The preview image is always garbled. I tried using the master and the staging releases and ran into the same issue. 

Any scan I make below 200 dpi is unusable. If I scan at 200, 300, or 600 dpi everything is fine. The file format I choose does not seem to matter, but I have not tried every resolution and every file format. 

I would think higher-dpi output would be the troublemaker, but it works fine. I tried doing some scans from the command line with scanimage, and all output was normal, regardless of resolution. 

Low-dpi (and preview) images look like this:
https://imgpile.com/i/7RzgUM

Client:
Brave Version 1.18.77 Chromium: 87.0.4280.101 (Official Build) (64-bit)
Firefox 84.0.1

Server:
OS: Ubuntu 20.04
Docker: Docker version 19.03.8
NPM & Node: Whatever the docker installer picks

Logs:
https://pastebin.com/cMjDWiqB

**Comment by sbs20** [comment]
Thanks for reporting. So just to restate the problem:

* The preview image is **always** garbled.
* But if you do a "scan" then it's only garbled below 200dpi, and 200dpi and above look fine.

Is that correct? I'm asking because the preview process does things slightly differently and it will help narrow down the problem.

Also:

>  I tried doing some scans from the command line with scanimage, and all output was normal, regardless of resolution.

Can you send the command that you used? I've seen some odd behaviour with scanimage where the parameter order seems to matter. That may also help us narrow down what's happening.

Scanimage is doing this:

```
/usr/bin/scanimage -d 'xerox_mfp:libusb:001:002' --mode 'Color' --source 'Flatbed' --resolution 100 -l 0 -t 0 -x 212 -y 291 --format 'tiff' > ~/test-0001.tif
/usr/bin/scanimage -d 'xerox_mfp:libusb:001:002' --mode 'Color' --source 'Flatbed' --resolution 200 -l 0 -t 0 -x 212 -y 291 --format 'tiff' > ~/test-0002.tif
```

If you run those commands hopefully we can reproduce that test-0001.tif is garbled and 0002 is not.

**Comment by sbs20** [comment]
Keeping [logs.txt](https://github.com/sbs20/scanservjs/files/5792647/logs.txt) for posterity

**Comment by WayneManion** [comment]
[url=https://postimg.cc/pyrkR8Bc][img]https://i.postimg.cc/pyrkR8Bc/test-0001.png[/img][/url]

[url=https://postimg.cc/mhnjDDXL][img]https://i.postimg.cc/mhnjDDXL/test-0002.png[/img][/url]

**Comment by sbs20** [comment]
Bizarre!

Ok can you try the following?

```
# Control experiment - remove geometry
/usr/bin/scanimage -d 'xerox_mfp:libusb:001:002' --mode 'Color' --source 'Flatbed' --resolution 100 --format 'tiff' > /home/test-0011.tif

# Remove source as per working example, but output tiff 
/usr/bin/scanimage -d 'xerox_mfp:libusb:001:002' --mode 'Color' --resolution 100 --format 'tiff' > /home/test-0012.tif

# Move source before mode
/usr/bin/scanimage -d 'xerox_mfp:libusb:001:002' --source 'Flatbed' --mode 'Color' --resolution 100 --format 'tiff' > /home/test-0013.tif

# As per scanserv but format as jpeg
/usr/bin/scanimage -d 'xerox_mfp:libusb:001:002' --mode 'Color' --source 'Flatbed' --resolution 100 --format 'jpeg' > /home/test-0014.jpg

# Put source after resolution
/usr/bin/scanimage -d 'xerox_mfp:libusb:001:002' --mode 'Color' --resolution 100 --source 'Flatbed' --format 'tiff' > /home/test-0015.tif
```

All seems weird but then [this](https://github.com/sbs20/scanservjs/issues/100#issuecomment-738141922) happened so who knows!?

**Comment by WayneManion** [comment]
I ran all of those commands and all resulted in good images.

**Comment by sbs20** [comment]
What??? So it seems to be the geometry then. What about these?

(Sorry about this - I just can't reproduce with my scanners. Thanks for your patience and help!)

```
/usr/bin/scanimage -d 'xerox_mfp:libusb:001:002' --mode 'Color' --source 'Flatbed' --resolution 100 -l 0 -t 0 --format 'tiff' > /home/test-0021.tif
/usr/bin/scanimage -d 'xerox_mfp:libusb:001:002' --mode 'Color' --source 'Flatbed' --resolution 100 -l 0 -t 0 -x 212 -y 291 --format 'tiff' > /home/test-0022.tif
/usr/bin/scanimage -d 'xerox_mfp:libusb:001:002' --mode 'Color' --source 'Flatbed' --resolution 100 -l 0 -t 0 -x 215 -y 297 --format 'tiff' > /home/test-0023.tif
```

**Comment by WayneManion** [comment]
test-0021: OK
/usr/bin/scanimage -d 'xerox_mfp:libusb:001:002' --mode 'Color' --source 'Flatbed' --resolution 100 -l 0 -t 0 --format 'tiff' > /home/test-0021.tif
https://i.postimg.cc/3r6TKKjk/test-0021.png

test-0022: garbled
/usr/bin/scanimage -d 'xerox_mfp:libusb:001:002' --mode 'Color' --source 'Flatbed' --resolution 100 -l 0 -t 0 -x 212 -y 291 --format 'tiff' > /home/test-0022.tif
https://i.postimg.cc/130hcwkF/test-0022.png

test-0023: garbled, but different "garble angle"
/usr/bin/scanimage -d 'xerox_mfp:libusb:001:002' --mode 'Color' --source 'Flatbed' --resolution 100 -l 0 -t 0 -x 215 -y 297 --format 'tiff' > /home/test-0023.tif
https://i.postimg.cc/N0ghNLtz/test-0023.png

**Comment by sbs20** [comment]
Ok. I need to have a think how to fix it. Thanks again for your help.

**Comment by WayneManion** [comment]
Thank you for making this neat tool in the first place.

**Comment by sbs20** [comment]
## TL;DR;
Can you provide the following? I think we need to search for more info on a fix with sane-utils. In the mean time see the notes on option 4 below.

```
uname -a
scanimage -V
sudo sane-find-scanner -v -v
```

## Notes for fix:

* The problem appears to be that inclusion of the `-x` and `-y` geometry causes odd image shearing for this scanner when the resolution is under 200dpi.
* This behaviour is duplicated at the command line - it's a scanimage / sane-utils issue (not specifically scanservjs)
* Not all scanners have the default -x and -y as the maximum values.

<details>
  <summary>Canon CanoScan N670U/N676U/LiDE20 example (click to expand)</summary>

```console
$ scanimage -A
All options specific to device `plustek:libusb:001:003':
  Scan Mode:
    --mode Lineart|Gray|Color [Color]
        Selects the scan mode (e.g., lineart, monochrome, or color).
    --depth 8|16bit [8]
        Number of bits per sample, typical values are 1 for "line-art" and 8
        for multibit scans.
    --source Normal|Transparency|Negative [inactive]
        Selects the scan source (such as a document-feeder).
    --resolution 50..1200dpi [50]
        Sets the resolution of the scanned image.
    --preview[=(yes|no)] [no]
        Request a preview-quality scan.
  Geometry:
    -l 0..215mm [0]
        Top-left x position of scan area.
    -t 0..297mm [0]
        Top-left y position of scan area.
    -x 0..215mm [103] <------- 103?
        Width of scan-area.
    -y 0..297mm [76.21] <------- 76.21?
        Height of scan-area.
```

  </details>

## Options:
1. Stay as-is: This bug remains. There is a workaround for actual scanning (use a higher resolution), but preview remains broken
2. Do not include geometry for preview: This would break the Canon case above.
3. Allow user to specify whether to include geometry - doesn't fix preview which has fixed DPI
4. Increase DPI for preview - slows it down creates more network traffic (although it's LAN and relatively low). Would also need to find a working DPI (not all scanners will support 200).
   This can be manually overridden for now - edit `Config.previewResolution = 200;` in `./config/Config.js`
5. Investigate [known defects with scanimage and Samsung (xerox) scanners](https://gitlab.com/sane-project/backends/-/issues?scope=all&utf8=%E2%9C%93&state=opened&search=xerox_mfp)

## Current resolution
* Options 1-3 are all bad.
* 4 is a possibility but is just a less bad workaround.
* 5 is best but more effort and slower.

**Comment by WayneManion** [comment]
Use of geometry could be selectable by the user? Maybe save that input across sessions?

**Comment by WayneManion** [comment]
Sorry, I missed the part asking me to run those three commands the first time I read your response. 
```
$ uname -a
Linux printerscanner 5.4.0-62-generic #70-Ubuntu SMP Tue Jan 12 12:45:47 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

$ scanimage -V
scanimage (sane-backends) 1.0.29; backend version 1.0.29

$ sudo sane-find-scanner -v -v
[sudo] password for wayne: 
This is sane-find-scanner from sane-backends 1.0.29

  # sane-find-scanner will now attempt to detect your scanner. If the
  # result is different from what you expected, first make sure your
  # scanner is powered up and properly connected to your computer.

searching for SCSI scanners:
checking /dev/scanner... failed to open (Invalid argument)
checking /dev/sg0... failed to open (Invalid argument)
checking /dev/sg1... failed to open (Invalid argument)
checking /dev/sg2... failed to open (Invalid argument)
checking /dev/sg3... failed to open (Invalid argument)
checking /dev/sg4... failed to open (Invalid argument)
checking /dev/sg5... failed to open (Invalid argument)
checking /dev/sg6... failed to open (Invalid argument)
checking /dev/sg7... failed to open (Invalid argument)
checking /dev/sg8... failed to open (Invalid argument)
checking /dev/sg9... failed to open (Invalid argument)
checking /dev/sga... failed to open (Invalid argument)
checking /dev/sgb... failed to open (Invalid argument)
checking /dev/sgc... failed to open (Invalid argument)
checking /dev/sgd... failed to open (Invalid argument)
checking /dev/sge... failed to open (Invalid argument)
checking /dev/sgf... failed to open (Invalid argument)
checking /dev/sgg... failed to open (Invalid argument)
checking /dev/sgh... failed to open (Invalid argument)
checking /dev/sgi... failed to open (Invalid argument)
checking /dev/sgj... failed to open (Invalid argument)
checking /dev/sgk... failed to open (Invalid argument)
checking /dev/sgl... failed to open (Invalid argument)
checking /dev/sgm... failed to open (Invalid argument)
checking /dev/sgn... failed to open (Invalid argument)
checking /dev/sgo... failed to open (Invalid argument)
checking /dev/sgp... failed to open (Invalid argument)
checking /dev/sgq... failed to open (Invalid argument)
checking /dev/sgr... failed to open (Invalid argument)
checking /dev/sgs... failed to open (Invalid argument)
checking /dev/sgt... failed to open (Invalid argument)
checking /dev/sgu... failed to open (Invalid argument)
checking /dev/sgv... failed to open (Invalid argument)
checking /dev/sgw... failed to open (Invalid argument)
checking /dev/sgx... failed to open (Invalid argument)
checking /dev/sgy... failed to open (Invalid argument)
checking /dev/sgz... failed to open (Invalid argument)
  # No SCSI scanners found. If you expected something different, make sure that
  # you have loaded a kernel SCSI driver for your SCSI adapter.

searching for USB scanners:
checking /dev/usb/scanner... failed to open (Invalid argument)
checking /dev/usb/scanner0... failed to open (Invalid argument)
checking /dev/usb/scanner1... failed to open (Invalid argument)
checking /dev/usb/scanner2... failed to open (Invalid argument)
checking /dev/usb/scanner3... failed to open (Invalid argument)
checking /dev/usb/scanner4... failed to open (Invalid argument)
checking /dev/usb/scanner5... failed to open (Invalid argument)
checking /dev/usb/scanner5... failed to open (Invalid argument)
checking /dev/usb/scanner7... failed to open (Invalid argument)
checking /dev/usb/scanner8... failed to open (Invalid argument)
checking /dev/usb/scanner9... failed to open (Invalid argument)
checking /dev/usb/scanner10... failed to open (Invalid argument)
checking /dev/usb/scanner11... failed to open (Invalid argument)
checking /dev/usb/scanner12... failed to open (Invalid argument)
checking /dev/usb/scanner13... failed to open (Invalid argument)
checking /dev/usb/scanner14... failed to open (Invalid argument)
checking /dev/usb/scanner15... failed to open (Invalid argument)
checking /dev/usbscanner... failed to open (Invalid argument)
checking /dev/usbscanner0... failed to open (Invalid argument)
checking /dev/usbscanner1... failed to open (Invalid argument)
checking /dev/usbscanner2... failed to open (Invalid argument)
checking /dev/usbscanner3... failed to open (Invalid argument)
checking /dev/usbscanner4... failed to open (Invalid argument)
checking /dev/usbscanner5... failed to open (Invalid argument)
checking /dev/usbscanner6... failed to open (Invalid argument)
checking /dev/usbscanner7... failed to open (Invalid argument)
checking /dev/usbscanner8... failed to open (Invalid argument)
checking /dev/usbscanner9... failed to open (Invalid argument)
checking /dev/usbscanner10... failed to open (Invalid argument)
checking /dev/usbscanner11... failed to open (Invalid argument)
checking /dev/usbscanner12... failed to open (Invalid argument)
checking /dev/usbscanner13... failed to open (Invalid argument)
checking /dev/usbscanner14... failed to open (Invalid argument)
checking /dev/usbscanner15... failed to open (Invalid argument)
trying libusb:

<device descriptor of 0x1d6b/0x0003 at 002:001 (Linux 5.4.0-62-generic xhci-hcd xHCI Host Controller)>
bLength               18
bDescriptorType       1
bcdUSB                3.00
bDeviceClass          9
bDeviceSubClass       0
bDeviceProtocol       3
bMaxPacketSize0       9
idVendor              0x1D6B
idProduct             0x0003
bcdDevice             5.04
iManufacturer         3 (Linux 5.4.0-62-generic xhci-hcd)
iProduct              2 (xHCI Host Controller)
iSerialNumber         1 (0000:00:14.0)
bNumConfigurations    1
 <configuration 0>
 bLength              9
 bDescriptorType      2
 wTotalLength         31
 bNumInterfaces       1
 bConfigurationValue  1
 iConfiguration       0 ()
 bmAttributes         224 (Self-poweredRemote Wakeup)
 MaxPower             0 mA
  <interface 0>
   <altsetting 0>
   bLength            9
   bDescriptorType    4
   bInterfaceNumber   0
   bAlternateSetting  0
   bNumEndpoints      1
   bInterfaceClass    9
   bInterfaceSubClass 0
   bInterfaceProtocol 0
   iInterface         0 ()
    <endpoint 0>
    bLength           7
    bDescriptorType   5
    bEndpointAddress  0x81 (in 0x01)
    bmAttributes      3 (interrupt)
    wMaxPacketSize    4
    bInterval         12 ms
    bRefresh          0
    bSynchAddress     0

<device descriptor of 0x1a86/0xe024 at 001:003>
bLength               18
bDescriptorType       1
bcdUSB                1.10
bDeviceClass          0
bDeviceSubClass       0
bDeviceProtocol       0
bMaxPacketSize0       64
idVendor              0x1A86
idProduct             0xE024
bcdDevice             1.012
iManufacturer         0 ()
iProduct              0 ()
iSerialNumber         0 ()
bNumConfigurations    1
 <configuration 0>
 bLength              9
 bDescriptorType      2
 wTotalLength         34
 bNumInterfaces       1
 bConfigurationValue  1
 iConfiguration       0 ()
 bmAttributes         128 ()
 MaxPower             100 mA
  <interface 0>
   <altsetting 0>
   bLength            9
   bDescriptorType    4
   bInterfaceNumber   0
   bAlternateSetting  0
   bNumEndpoints      1
   bInterfaceClass    3
   bInterfaceSubClass 0
   bInterfaceProtocol 0
   iInterface         0 ()
    <endpoint 0>
    bLength           7
#    bDescriptorType   5
    bEndpointAddress  0x82 (in 0x02)
    bmAttributes      3 (interrupt)
    wMaxPacketSize    64
    bInterval         1 ms
    bRefresh          0
    bSynchAddress     0

<device descriptor of 0x04e8/0x3434 at 001:002 (Samsung SCX-4623 Series)>
bLength               18
bDescriptorType       1
bcdUSB                2.00
bDeviceClass          0
bDeviceSubClass       0
bDeviceProtocol       0
bMaxPacketSize0       64
idVendor              0x04E8
idProduct             0x3434
bcdDevice             1.00
iManufacturer         1 (Samsung)
iProduct              2 (SCX-4623 Series)
iSerialNumber         3 (Z2TDBABB702479W)
bNumConfigurations    1
 <configuration 0>
 bLength              9
 bDescriptorType      2
 wTotalLength         55
 bNumInterfaces       2
 bConfigurationValue  1
 iConfiguration       0 ()
 bmAttributes         192 (Self-powered)
 MaxPower             2 mA
  <interface 0>
   <altsetting 0>
   bLength            9
   bDescriptorType    4
   bInterfaceNumber   0
   bAlternateSetting  0
   bNumEndpoints      2
   bInterfaceClass    255
   bInterfaceSubClass 255
   bInterfaceProtocol 255
   iInterface         0 ()
    <endpoint 0>
    bLength           7
    bDescriptorType   5
    bEndpointAddress  0x03 (out 0x03)
    bmAttributes      2 (bulk)
    wMaxPacketSize    512
    bInterval         10 ms
    bRefresh          0
    bSynchAddress     0
    <endpoint 1>
    bLength           7
    bDescriptorType   5
    bEndpointAddress  0x84 (in 0x04)
    bmAttributes      2 (bulk)
    wMaxPacketSize    512
    bInterval         10 ms
    bRefresh          0
    bSynchAddress     0
  <interface 1>
   <altsetting 0>
   bLength            9
   bDescriptorType    4
   bInterfaceNumber   1
   bAlternateSetting  0
   bNumEndpoints      2
   bInterfaceClass    7
   bInterfaceSubClass 1
   bInterfaceProtocol 2
   iInterface         0 ()
    <endpoint 0>
    bLength           7
    bDescriptorType   5
    bEndpointAddress  0x01 (out 0x01)
    bmAttributes      2 (bulk)
    wMaxPacketSize    512
    bInterval         10 ms
    bRefresh          0
    bSynchAddress     0
    <endpoint 1>
    bLength           7
    bDescriptorType   5
    bEndpointAddress  0x82 (in 0x02)
    bmAttributes      2 (bulk)
    wMaxPacketSize    512
    bInterval         10 ms
    bRefresh          0
    bSynchAddress     0

<trying to find out which USB chip is used>
    checking for LM983[1,2,3] ...
    this is not a LM983x (bcdUSB = 0x200)
    checking for GT-6801 ...
    this is not a GT-6801 (bDeviceClass = 0)
    checking for GT-6816 ...
    this is not a GT-6816 (bcdUSB = 0x200)
    checking for GLxxx ...
    this is not a GL646 (bDeviceClass = 0, bInterfaceClass = 255)
    this is not a GLxxx (bDeviceClass = 0, bInterfaceClass = 255)
<Couldn't determine the type of the USB chip (result from sane-backends 1.0.29)>

found USB scanner (vendor=0x04e8 [Samsung], product=0x3434 [SCX-4623 Series]) at libusb:001:002

<device descriptor of 0x1d6b/0x0002 at 001:001 (Linux 5.4.0-62-generic xhci-hcd xHCI Host Controller)>
bLength               18
bDescriptorType       1
bcdUSB                2.00
bDeviceClass          9
bDeviceSubClass       0
bDeviceProtocol       1
bMaxPacketSize0       64
idVendor              0x1D6B
idProduct             0x0002
bcdDevice             5.04
iManufacturer         3 (Linux 5.4.0-62-generic xhci-hcd)
iProduct              2 (xHCI Host Controller)
iSerialNumber         1 (0000:00:14.0)
bNumConfigurations    1
 <configuration 0>
 bLength              9
 bDescriptorType      2
 wTotalLength         25
 bNumInterfaces       1
 bConfigurationValue  1
 iConfiguration       0 ()
 bmAttributes         224 (Self-poweredRemote Wakeup)
 MaxPower             0 mA
  <interface 0>
   <altsetting 0>
   bLength            9
   bDescriptorType    4
   bInterfaceNumber   0
   bAlternateSetting  0
   bNumEndpoints      1
   bInterfaceClass    9
   bInterfaceSubClass 0
   bInterfaceProtocol 0
   iInterface         0 ()
    <endpoint 0>
    bLength           7
    bDescriptorType   5
    bEndpointAddress  0x81 (in 0x01)
    bmAttributes      3 (interrupt)
    wMaxPacketSize    4
    bInterval         12 ms
    bRefresh          0
    bSynchAddress     0
  # Your USB scanner was (probably) detected. It may or may not be supported by
  # SANE. Try scanimage -L and read the backend's manpage.

  # Not checking for parallel port scanners.

  # Most Scanners connected to the parallel port or other proprietary ports
  # can't be detected by this program.
done

```

**Comment by sbs20** [comment]
I have made the preview override a bit easier. See [here](https://github.com/sbs20/scanservjs#configuration-override) for an example. I will keep this issue open a bit longer and will attempt to get around to filing an issue with the scanimage crew to see if they know what's up.

**Comment by sbs20** [comment]
I've been giving this quite a lot of thought. Saving the use of geometry is going to be a reasonably large change - it's front and back end changes and will result in odd behaviour for some users (e.g. the Canon case above). If there weren't a workaround I would be more inclined to do it.

If I were able to test it myself I would also experiment with newer build of SANE and scanimage to see if the issue has been fixed since `1.0.29`- it takes quite a while for new builds to filter through all the distros, especially Debian / Ubuntu.

I believe that adding the following `config/config.local.js` file will resolve this issue for you. It's an imperfect solution, but I think reasonable.

```javascript
module.exports = {
  /**
   * @param {Configuration} config 
   */
  afterConfig(config) {
    // Set default preview resolution
    config.previewResolution = 200;
  },

  /**
   * @param {ScanDevice[]} devices 
   */
  afterDevices(devices) {
    // Override the defaults for plustek scanners
    const device = devices.filter(d => d.id.startsWith('xerox_mfp'))[0];
    if (device) {
      device.features['--resolution'].default = 200;
      device.features['--resolution'].options = [200, 300, 600];
    }
  }
};
```

I'll close this issue for now - but do feel free to re-open if you have new information to hand. I would like to resolve this if possible.
